File: frontend/lib/core/models/account.dart
/// 本地账户模型（与后端 sync 协议对齐，uuid 为跨端身份键）。

class LocalAccount {
  final String uuid;
  final String name; // Beancount ASCII 账户名
  final String? displayName; // 中文显示名（可选）
  final String type; // Assets / Liabilities / Equity / Income / Expenses
  final String openDate;
  final String? closeDate; // 关闭日（YYYY-MM-DD）；非空表示已关闭（Beancount close，0.4-A）
  final String? restriction; // 限定币种（可选）
  final String? icon; // 图标（emoji，展示用）
  final String? color; // 颜色 hex（展示用）
  final String? parentUuid; // 父账户 uuid（三级分类）
  final String? subType; // 账户子类型（展示用）
  final int sortOrder; // 用户自定义排序（0.4-C）；同类型内升序，0 表示未设置

  LocalAccount({
    required this.uuid,
    required this.name,
    this.displayName,
    required this.type,
    required this.openDate,
    this.closeDate,
    this.restriction,
    this.icon,
    this.color,
    this.parentUuid,
    this.subType,
    this.sortOrder = 0,
  });

  /// 是否已关闭（close_date 非空）。
  bool get isClosed => closeDate != null && closeDate!.isNotEmpty;

  /// 展示名：优先中文 display_name，回退 ASCII name。
  String get display => (displayName?.isNotEmpty == true) ? displayName! : name;

  Map<String, dynamic> toMap() => {
        'uuid': uuid,
        'name': name,
        'display_name': displayName,
        'type': type,
        'open_date': openDate,
        'close_date': closeDate,
        'commodity_restriction': restriction,
        'icon': icon,
        'color': color,
        'parent_uuid': parentUuid,
        'sub_type': subType,
        'sort_order': sortOrder,
      };

  factory LocalAccount.fromMap(Map<String, dynamic> m) => LocalAccount(
        uuid: m['uuid'] as String,
        name: m['name'] as String,
        displayName: m['display_name'] as String?,
        type: m['type'] as String,
        openDate: m['open_date'] as String,
        closeDate: m['close_date'] as String?,
        restriction: m['commodity_restriction'] as String?,
        icon: m['icon'] as String?,
        color: m['color'] as String?,
        parentUuid: m['parent_uuid'] as String?,
        subType: m['sub_type'] as String?,
        sortOrder: m['sort_order'] as int? ?? 0,
      );
}

/// 账户展示名：优先中文 display_name，回退 ASCII name。
String accountDisplayName(LocalAccount a) => a.display;

File: internal/service/account.go
package service

import (
	"time"

	"familyledger/internal/domain"

	"github.com/google/uuid"
)

// AccountView 是账户列表项的返回视图（含按币种余额）。
type AccountView struct {
	ID          int64             `json:"id"`
	Name        string            `json:"name"`
	DisplayName string            `json:"display_name,omitempty"`
	Type        string            `json:"type"`
	OpenDate    string            `json:"open_date"`
	CloseDate   string            `json:"close_date,omitempty"` // 非空表示已关闭（Beancount close）
	Restriction string            `json:"commodity_restriction,omitempty"`
	Icon        string            `json:"icon,omitempty"`
	Color       string            `json:"color,omitempty"`
	ParentUUID  string            `json:"parent_uuid,omitempty"`
	SubType     string            `json:"sub_type,omitempty"`
	SortOrder   int64             `json:"sort_order"` // 0.4-C 分类拖拽排序（同类型内升序）
	Balances    map[string]string `json:"balances,omitempty"`
}

// AccountCreated 是创建账户的返回。
type AccountCreated struct {
	ID          int64  `json:"id"`
	UUID        string `json:"uuid"`
	Name        string `json:"name"`
	DisplayName string `json:"display_name,omitempty"`
	Type        string `json:"type"`
	OpenDate    string `json:"open_date"`
	Icon        string `json:"icon,omitempty"`
	Color       string `json:"color,omitempty"`
	ParentUUID  string `json:"parent_uuid,omitempty"`
	SubType     string `json:"sub_type,omitempty"`
}

// ListAccounts 列出账本内账户及其按币种余额。includeClosed=false 仅未关闭账户。
func (s *Service) ListAccounts(ledgerID int64, includeClosed bool) ([]AccountView, error) {
	accounts, err := s.repo.ListAccounts(ledgerID, includeClosed)
	if err != nil {
		return nil, err
	}
	balances, err := s.repo.ComputeBalances(ledgerID)
	if err != nil {
		return nil, err
	}

	out := make([]AccountView, 0, len(accounts))
	for _, a := range accounts {
		out = append(out, AccountView{
			ID:          a.ID,
			Name:        a.Name,
			DisplayName: a.DisplayName,
			Type:        a.Type,
			OpenDate:    a.OpenDate,
			CloseDate:   a.CloseDate,
			Restriction: a.Restriction,
			Icon:        a.Icon,
			Color:       a.Color,
			ParentUUID:  a.ParentUUID,
			SubType:     a.SubType,
			SortOrder:   a.SortOrder,
			Balances:    balances[a.ID],
		})
	}
	return out, nil
}

// CreateAccount 创建账户（含币种限制时自动补建币种）。
// icon/color/parentUUID/subType 为展示/层级字段，可选。
func (s *Service) CreateAccount(ledgerID int64, name, displayName, atype, openDate, commodity, icon, color, parentUUID, subType string) (AccountCreated, error) {
	if name == "" || atype == "" {
		return AccountCreated{}, domain.Invalidf("name 与 type 必填")
	}
	if err := validateAccountType(atype); err != nil {
		return AccountCreated{}, err
	}
	if openDate == "" {
		openDate = time.Now().Format("2006-01-02")
	}

	a := domain.Account{
		UUID:        uuid.NewString(),
		Name:        name,
		DisplayName: displayName,
		Type:        atype,
		OpenDate:    openDate,
		Restriction: commodity,
		Icon:        icon,
		Color:       color,
		ParentUUID:  parentUUID,
		SubType:     subType,
	}
	id, err := s.repo.InsertAccount(ledgerID, a)
	if err != nil {
		return AccountCreated{}, err
	}
	return AccountCreated{
		ID: id, UUID: a.UUID, Name: name, DisplayName: displayName, Type: atype,
		OpenDate: openDate, Icon: icon, Color: color, ParentUUID: parentUUID, SubType: subType,
	}, nil
}

// DeleteAccount 关闭账户（Beancount close 语义，0.4-A 起）。
// 置 close_date，保留全部交易引用；sync_log 写 op=close。
// HTTP 层仍以 DELETE /api/accounts/{uuid} 暴露，语义从软删改为关闭。
func (s *Service) DeleteAccount(ledgerID int64, uuid string) error {
	closed, err := s.repo.CloseAccount(ledgerID, uuid)
	if err != nil {
		return err
	}
	if !closed {
		return domain.ErrNotFound
	}
	return nil
}
